Lib/pathlib/_os.py

```diff
@@ -166,48 +166,78 @@ def copyfileobj(source_f, target_f):
         write_target(buf)
 
 
-def magic_open(path, mode='r', buffering=-1, encoding=None, errors=None,
-               newline=None):
+def _open_reader(obj):
+    cls = type(obj)
+    try:
+        return cls.__open_reader__(obj)
+    except AttributeError:
+        if hasattr(cls, '__open_reader__'):
+            raise
+    raise TypeError(f"{cls.__name__} can't be opened for reading")
+
+
+def _open_writer(obj, mode):
+    cls = type(obj)
+    try:
+        return cls.__open_writer__(obj, mode)
+    except AttributeError:
+        if hasattr(cls, '__open_writer__'):
+            raise
+    raise TypeError(f"{cls.__name__} can't be opened for writing")
+
+
+def _open_updater(obj, mode):
+    cls = type(obj)
+    try:
+        return cls.__open_updater__(obj, mode)
+    except AttributeError:
+        if hasattr(cls, '__open_updater__'):
+            raise
+    raise TypeError(f"{cls.__name__} can't be opened for updating")
+
+
+def vfsopen(obj, mode='r', buffering=-1, encoding=None, errors=None,
+            newline=None):
     """
     Open the file pointed to by this path and return a file object, as
     the built-in open() function does.
+
+    Unlike the built-in open() function, this function accepts 'openable'
+    objects, which are objects with any of these magic methods:
```

Comment:
```suggestion
    objects, which are objects with any of these special methods:
```
See [the glossary](https://docs.python.org/3/glossary.html#term-special-method).